app/src/main/java/com/jerboa/model/PostsViewModel.kt

```diff
@@ -18,40 +18,37 @@ import com.jerboa.api.toApiState
 import com.jerboa.datatypes.BanFromCommunityData
 import com.jerboa.db.entity.AnonAccount
 import com.jerboa.db.repository.AccountRepository
-import com.jerboa.findAndUpdatePost
-import com.jerboa.findAndUpdatePostCreator
-import com.jerboa.findAndUpdatePostCreatorBannedFromCommunity
+import com.jerboa.feed.PaginationController
+import com.jerboa.feed.PostController
 import com.jerboa.findAndUpdatePostHidden
-import com.jerboa.mergePosts
 import com.jerboa.toEnumSafe
 import it.vercruysse.lemmyapi.dto.ListingType
 import it.vercruysse.lemmyapi.dto.SortType
 import it.vercruysse.lemmyapi.v0x19.datatypes.CreatePostLike
 import it.vercruysse.lemmyapi.v0x19.datatypes.DeletePost
 import it.vercruysse.lemmyapi.v0x19.datatypes.FeaturePost
 import it.vercruysse.lemmyapi.v0x19.datatypes.GetPosts
-import it.vercruysse.lemmyapi.v0x19.datatypes.GetPostsResponse
 import it.vercruysse.lemmyapi.v0x19.datatypes.HidePost
 import it.vercruysse.lemmyapi.v0x19.datatypes.LockPost
 import it.vercruysse.lemmyapi.v0x19.datatypes.MarkPostAsRead
-import it.vercruysse.lemmyapi.v0x19.datatypes.PaginationCursor
 import it.vercruysse.lemmyapi.v0x19.datatypes.PersonView
 import it.vercruysse.lemmyapi.v0x19.datatypes.PostView
 import it.vercruysse.lemmyapi.v0x19.datatypes.SavePost
 import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.launch
 
 open class PostsViewModel(protected val accountRepository: AccountRepository) : ViewModel() {
-    var postsRes: ApiState<GetPostsResponse> by mutableStateOf(ApiState.Empty)
-        private set
-    private var page by mutableLongStateOf(1)
-    protected var pageCursor: PaginationCursor? by mutableStateOf(null)
+    val lazyListState = LazyListState()
+    var postsRes: ApiState<List<PostView>> by mutableStateOf(ApiState.Empty)
```

Comment:
This one still needs to be got rid of then I think. Sometimes below you are updating the postController, and sometimes the postsRes. Lets try to keep only one list.

Reply:
We need to update both PostRes and the PostController. PostRes still controls the api states, ex: loading/appending show loading bar. PostController controls the list in the postRes

Reply:
We store one list, the other is just a reference to it.

Doing `ApiState<SnapshotStateList<PostView>>` would me we expose the `SnapshotStateList` everywhere so everywhere we could update the apistate from. 

`List<PostView>` is immutable. So can't be manipulated.

Reply:
L75 has `postController.addAll(it.posts)` , adding all the items individually to a separate list.

L96-97 Also add items to both lists, twice.

Reply:
I think this was the point of the `private set` also, to ensure that its source data can only change in the model.

I need to read more about snapshotstatelist, because this is a problem that should be able to be solved without keeping two lists.

Reply:
L75 is the initPosts, on startup we load the first page. So we add all the items.

On L96-99 is appendPosts (load next page). so we add all the new posts

Ig they are kinda the same and I could just use appendPosts but this is how we do it everywhere. (init then append for next pages)

What am I missing?

Reply:
`private set` doesn't prevent us from manipulating the inner state of apiState just prevent us from reassigning it to another.

Reply:
The two lists are the `SnapshotStateList` in FeedController, and the `List` wrapped in the ApiState in `postsRes`.

> Doing ApiState<SnapshotStateList<PostView>> would mean we expose the SnapshotStateList everywhere, so everywhere we could update the posts from.

I spose this is all a workaround due to the fact that we can't make `SnapshotStateList` externally safe, although to me it'd be logically cleaner to just do `ApiState<SnapshotStateList<PostView>>`.

Did you test to make sure things like `postController.findAndUpdateCreator(personView.person)` for banning / unbanning work correctly? To me it seems like these aren't going to work correctly, since externally only `postsRes` is exposed, not your new snapshotstatelist.

Also for example on L97, you're doing `postsRes = ApiState.Success(postController.feed)`, which to me seems like its going to replace the list anyway, so you'd lose all the benefits of the snapshotstatelist.

Reply:
There is only one list, `SnapshotStateList ` in FeedController. I expose this as `List` in the feed controller. All operations are done on this (SnapShotState)List. During appending I change the `ApiState` to  appending ` ApiState.Appending(oldRes.data)` So that the loading bar, etc trigggers but its still the same list.

then when I receive the newPosts from the API. I add the posts to the `SnapShotStateList`
`postController.addAll(newRes.data.posts)` then I change ApiState back to Success ` postsRes = ApiState.Success(postController.feed)` so that the loading bar goes away.

I could do `postsRes = ApiState.Success(oldRes.data)` same thing. There is only ever one list and all contain a reference to it.


> Did you test to make sure things like (...)

Yes it works see vid below. Only `postRes` is indeed exposed and `postRes` exposes that `SnapShotStateList` as `List`

Reply:
Alright, these cases are a bit confusing to me, but that's my bad.